processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,496 @@
+       */
+      private int remaining()
+      {
+        if (curWriteIndex >= nextReadIndex) {
+          return curWriteIndex - nextReadIndex;
+        } else {
+          return (maxNumSlots - nextReadIndex) + curWriteIndex;
+        }
+      }
+
+      @Override
+      public Entry<KeyType> next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        // Here, nextReadIndex should be valid which means:
+        // - a valid array index which should be >= 0 and < maxNumSlots
+        // - an index of the array slot where the aggregation for the corresponding grouping key is done
+        // - an index of the array slot which is not read yet
+        final int recordOffset = recordSize * nextReadIndex;
+        final KeyType key = keySerde.fromByteBuffer(buffer, recordOffset);
+
+        final Object[] values = new Object[aggregators.length];
+        for (int i = 0; i < aggregators.length; i++) {
+          values[i] = aggregators[i].get(buffer, recordOffset + aggregatorOffsets[i]);
+        }
+
+        final int targetIndex = nextReadIndex == maxNumSlots - 1 ? 0 : nextReadIndex + 1;
+        // Wait if the array is empty until at least one slot becomes available for read, and then increase
+        // nextReadIndex.
+        increaseReadIndexTo(targetIndex);
+
+        return new Entry<>(key, values);
+      }
+
+      /**
+       * Wait for {@link StreamingMergeSortedGrouper#curWriteIndex} to be moved if necessary and move
+       * {@link StreamingMergeSortedGrouper#nextReadIndex}.
+       *
+       * @param target the target index {@link StreamingMergeSortedGrouper#nextReadIndex} will move to
+       */
+      private void increaseReadIndexTo(int target)
+      {
+        // Check that the array is empty and wait for at least one slot to become available.
+        //
+        // curWriteIndex is a volatile variable and the changes on it are continuously checked until they are seen in
+        // waitFor(). See the following links.
+        // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.3.1.4
+        // * http://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5
+        // * https://stackoverflow.com/questions/11761552/detailed-semantics-of-volatile-regarding-timeliness-of-visibility
+        waitFor(
+            notUsed -> (curWriteIndex == -1 || target == curWriteIndex) &&
+               !finished && !Thread.currentThread().isInterrupted(),
+            queryTimeoutAtNs
+        );
+
+        // Changes on curWriteIndex happens-before changing nextReadIndex.
+        nextReadIndex = target;
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        // do nothing
+      }
+    };
+  }
+
+  /**
+   * Wait while the given predicate is true.  Throws an exception the current nano time is greater than the given
+   * timeout.
+   */
+  private static void waitFor(Predicate<Void> predicate, long queryTimeoutAtNs)
```

Comment:
Better create nested private interface called `Predicate` or `Condition` with a single no-arg boolean-returning method

Reply:
Removed this method.